package/MDAnalysis/coordinates/DCD.py

```diff
@@ -89,438 +69,214 @@
+        self._frame = -1
+        self._file.close()
+        self._file.open('r')
+
+    def _read_frame(self, i):
+        """read frame i"""
+        self._frame = i - 1
+        self._file.seek(i)
+        return self._read_next_timestep()
 
-    """
-    format = 'DCD'
-    flavor = 'CHARMM'
-    units = {'time': 'AKMA', 'length': 'Angstrom'}
-    _Timestep = Timestep
+    def _read_next_timestep(self, ts=None):
+        """copy next frame into timestep"""
+        if self._frame == self.n_frames - 1:
+            raise IOError('trying to go over trajectory limit')
+        if ts is None:
+            # use a copy to avoid that ts always points to the same reference
+            ts = self.ts.copy()
+        frame = self._file.read()
+        self._frame += 1
+        ts = self._frame_to_ts(frame, ts)
+        self.ts = ts
+        return ts
 
-    def __init__(self, dcdfilename, **kwargs):
-        super(DCDReader, self).__init__(dcdfilename, **kwargs)
+    def Writer(self, filename, n_atoms=None, **kwargs):
+        """Return writer for trajectory format"""
+        if n_atoms is None:
+            n_atoms = self.n_atoms
+        return DCDWriter(
+            filename,
+            n_atoms=n_atoms,
+            dt=self.ts.dt,
+            convert_units=self.convert_units,
+            **kwargs)
+
+    def _frame_to_ts(self, frame, ts):
+        """convert a dcd-frame to a mda TimeStep"""
+        ts.frame = self._frame
+        ts.time = ts.frame * self.ts.dt
+        ts.data['step'] = self._file.tell()
+
+        unitcell = frame.unitcell
+        pi_2 = np.pi / 2
+        if (-1.0 <= unitcell[1] <= 1.0) and (-1.0 <= unitcell[3] <= 1.0) and (
+                -1.0 <= unitcell[4] <= 1.0):
+            # This file was generated by Charmm, or by NAMD > 2.5, with the angle
+            # cosines of the periodic cell angles written to the DCD file.
+            # This formulation improves rounding behavior for orthogonal cells
+            # so that the angles end up at precisely 90 degrees, unlike acos().
+            # (changed in MDAnalysis 0.9.0 to have NAMD ordering of the angles;
+            # see Issue 187) */
+            alpha = 90.0 - np.arcsin(unitcell[4]) * 90.0 / pi_2
+            beta = 90.0 - np.arcsin(unitcell[3]) * 90.0 / pi_2
+            gamma = 90.0 - np.arcsin(unitcell[1]) * 90.0 / pi_2
+        else:
+            # This file was likely generated by NAMD 2.5 and the periodic cell
+            # angles are specified in degrees rather than angle cosines.
+            alpha = unitcell[4]
+            beta = unitcell[3]
+            gamma = unitcell[1]
+
+        unitcell[4] = alpha
+        unitcell[3] = beta
+        unitcell[1] = gamma
+
+        # The original unitcell is read as ``[A, gamma, B, beta, alpha, C]``
+        _ts_order = [0, 2, 5, 4, 3, 1]
+        uc = np.take(unitcell, _ts_order)
+        # heuristic sanity check: uc = A,B,C,alpha,beta,gamma
+        if np.any(uc < 0.) or np.any(uc[3:] > 180.):
```

Comment:
The way it is at the moment is technically wrong. Instead a better approach is
```python
if (-1.0 <= unitcell[1] <= 1.0) and (-1.0 <= unitcell[3] <= 1.0) and (-1.0 <= unitcell[4] <= 1.0):
   # do the angle cosines
elif np.any(unitcell < 0.) or np.any(unitcell[[1, 3, 4]] > 180.):
   # do the box vectors
else:
   # assume ABC angles
```

This is not worse than before and arguably more correct. I also don't see if this would miss angle cosines, given that that is the first `if` statement. Equally, if the second one triggers, it is surely is guaranteed not be lengths and angles.